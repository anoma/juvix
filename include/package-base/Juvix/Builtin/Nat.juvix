module Juvix.Builtin.Nat;

--- Inductive natural numbers. I.e. whole non-negative numbers.
builtin nat
type Nat :=
  | zero
  | suc Nat;
