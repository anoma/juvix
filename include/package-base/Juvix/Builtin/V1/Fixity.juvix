module Juvix.Builtin.V1.Fixity;

syntax fixity none := none;

syntax fixity rapp := binary {assoc := right};
syntax fixity lapp := binary {assoc := left; same := rapp};
syntax fixity seq := binary {assoc := left; above := [lapp]};

syntax fixity functor := binary {assoc := right};

syntax fixity logical := binary {assoc := right; above := [seq]};
syntax fixity comparison := binary {assoc := none; above := [logical]};

syntax fixity pair := binary {assoc := right};
syntax fixity cons := binary {assoc := right; above := [pair]};

syntax fixity step := binary {assoc := right};
syntax fixity range := binary {assoc := right; above := [step]};

syntax fixity additive := binary {assoc := left; above := [comparison; range; cons]};
syntax fixity multiplicative := binary {assoc := left; above := [additive]};

syntax fixity composition := binary {assoc := right; above := [multiplicative]};
syntax fixity lcomposition := binary {assoc := left; above := [multiplicative]};
