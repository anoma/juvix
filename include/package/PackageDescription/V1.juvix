module PackageDescription.V1;

import Stdlib.Prelude open;

--- A ;Package; defines the configuration for a Juvix package
type Package :=
  mkPackage {
    -- The name of the package
    name : String;
    -- The version for the package
    version : SemVer;
    -- The dependencies of this package
    dependencies : List Dependency;
    -- A path to the Main module for this package
    main : Maybe String;
    -- A path to a directory where Juvix should output intermediate build products
    buildDir : Maybe String
  };

open Package using {mkPackage} public;

--- A ;SemVer; defines a version that conforms to the Semantic Versioning specification.
type SemVer :=
  mkSemVer {
    major : Nat;
    minor : Nat;
    patch : Nat;
    release : Maybe String;
    meta : Maybe String
  };

open SemVer using {mkSemVer} public;

--- A ;Dependency; defines a Juvix package dependency.
type Dependency :=
  | --- A filesystem path to another Juvix Package.
    path {path : String}
  | --- A git repository containing a Juvix package at its root
    git {
      -- A name for this dependency
      name : String;
      -- The URL to the git repository
      url : String;
      -- The git ref to checkout
      ref : String
    }
  | --- The default Stdlib that is bundled with the Juvix compiler.
    defaultStdlib;

open Dependency using {path; git; defaultStdlib} public;

--- The default version used in `defaultPackage`.
defaultVersion : SemVer := mkVersion 0 0 0;

--- Construct a ;Package; with useful default arguments.
defaultPackage
  {name : String := "my-project"}
  {version : SemVer := defaultVersion}
  {dependencies : List Dependency := [defaultStdlib]}
  {main : Maybe String := nothing}
  {buildDir : Maybe String := nothing}
  : Package :=
  mkPackage name version dependencies main buildDir;

--- Construct a ;SemVer; with useful default arguments.
mkVersion
  (major minor patch : Nat)
  {release : Maybe String := nothing}
  {meta : Maybe String := nothing}
  : SemVer :=
  mkSemVer@{
    major;
    minor;
    patch;
    release;
    meta;
  };

--- Constructs a GitHub dependency.
github (org repo ref : String) : Dependency :=
  git
    (org ++str "_" ++str repo)
    ("https://github.com/" ++str org ++str "/" ++str repo)
    ref;
