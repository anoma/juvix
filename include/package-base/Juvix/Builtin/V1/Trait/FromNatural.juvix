module Juvix.Builtin.V1.Trait.FromNatural;

import Juvix.Builtin.V1.Nat.Base open using {Nat};

trait
type FromNatural A :=
  mkFromNatural {
    builtin from-nat
    fromNat : Nat -> A
  };

open FromNatural public;
