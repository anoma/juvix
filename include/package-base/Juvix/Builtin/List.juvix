module Juvix.Builtin.List;

import Stdlib.Data.Fixity open;

syntax operator :: cons;
--- Inductive list.
builtin list
type List (a : Type) :=
  | --- The empty list
    nil
  | --- An element followed by a list
    :: a (List a);
