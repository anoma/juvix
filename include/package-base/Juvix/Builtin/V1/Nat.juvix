module Juvix.Builtin.V1.Nat;

import Juvix.Builtin.V1.Trait.Natural open public;
import Juvix.Builtin.V1.Nat.Base open hiding {+; *; div; mod} public;
import Juvix.Builtin.V1.Nat.Base as Nat;

{-# specialize: true, inline: case #-}
instance
naturalNatI : Natural Nat :=
  mkNatural@{
    + := (Nat.+);
    * := (Nat.*);
    div := Nat.div;
    mod := Nat.mod;
    fromNat (x : Nat) : Nat := x
  };
