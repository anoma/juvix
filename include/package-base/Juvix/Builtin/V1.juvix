module Juvix.Builtin.V1;

import Juvix.Builtin.V1.Nat open public;
import Juvix.Builtin.V1.Trait.Natural open public;
import Juvix.Builtin.V1.String open public;
import Juvix.Builtin.V1.Bool open public;
import Juvix.Builtin.V1.Maybe open public;
import Juvix.Builtin.V1.List open public;
import Juvix.Builtin.V1.Fixity open public;
