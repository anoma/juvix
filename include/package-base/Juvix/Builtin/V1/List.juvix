module Juvix.Builtin.V1.List;

import Juvix.Builtin.V1.Fixity open;

syntax operator List.:: cons;
--- Inductive list.
builtin list
type List (A : Type) :=
  | --- The empty list
    nil
  | --- An element followed by a list
    :: A (List A);

open List using {nil; ::} public;
