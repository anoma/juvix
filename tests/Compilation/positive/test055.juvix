-- constructor printing
module test055;

import Juvix.Builtin.V1.Fixity open;
import Juvix.Builtin.V1.Trait.FromNatural open public;
import Juvix.Builtin.V1.Nat open;
import Juvix.Builtin.V1.List open;

type Pair' := pair : Nat -> Nat -> Pair';

open Pair' using {pair} public;

syntax operator Pair., pair;

--- Inductive pair. I.e. a tuple with two components.
builtin pair
type Pair (A B : Type) := , : A → B → Pair A B;

open Pair using {,} public;

main : Pair (List (Pair Pair' Nat)) (List Pair') :=
  (pair 1 2, 3) :: (pair 2 3, 4) :: nil, pair 1 2 :: pair 2 3 :: nil;
