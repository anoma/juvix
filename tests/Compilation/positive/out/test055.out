(pair 1 2, 3) :: (pair 2 3, 4) :: nil, pair 1 2 :: pair 2 3 :: nil
